core/utilities/ast.lua

```diff
@@ -1,13 +1,58 @@
---- SILE AST utilities
---
+--- AST utilities.
+-- Functions for working with SILE's Abstract Syntax Trees.
+-- @module SU.ast
+
+-- @type SU.ast
 local ast = {}
 
+--- Output developer friendly debugging view of an AST.
+-- @tparam table tree Abstract Syntax Tree.
+-- @tparam integer level Starting level to review.
+function ast.debug (tree, level)
+  if not tree then
+    SU.error("debugAST called with nil", true)
+  end
+  local out = string.rep("  ", 1+level)
+  if level == 0 then
+    SU.debug("ast", function ()
+      return "[" .. SILE.currentlyProcessingFile
+    end)
+  end
+  if type(tree) == "function" then
+    SU.debug("ast", function ()
+      return out .. tostring(tree)
+    end)
+  elseif type(tree) == "table" then
+    for _, content in ipairs(tree) do
+      if type(content) == "string" then
+        SU.debug("ast", function ()
+          return out .. "[" .. content .. "]"
+        end)
+      elseif type(content) == "table" then
+        if SILE.Commands[content.command] then
+          SU.debug("ast", function ()
+            return out .. "\\" .. content.command .. " " .. pl.pretty.write(content.options, "")
+          end)
+          if (#content>=1) then ast.debug(content, level+1) end
+        elseif content.id == "content" or (not content.command and not content.id) then
+          ast.debug(content, level+1)
+        else
+          SU.debug("ast", function ()
+            return out .. "?\\" .. (content.command or content.id)
+          end)
+        end
+      end
+    end
+  end
+  if level == 0 then SU.debug("ast", "]") end
+end
+
 --- Find a command node in a SILE AST tree,
---- looking only at the first level.
---- (We're not reimplementing XPath here.)
----@param tree      table       AST tree
----@param command   string      command name
----@return          table|nil   AST command node
+-- looking only at the first level.
+-- (We're not reimplementing XPath here.)
+-- @tparam table tree AST tree
```

Comment:
@Omikhleia I think I recall you saying the type hinds you added here worked in VSCode. I can't find the discussion but I have a memory of saying we probably wanted to use LDoc format in the future but that something was better than nothing to start with. This PR adds LDoc support and hence I tweaked these with type hints for that system. That being said it would be nice to figure out what else these can be used for and if we can figure out a way they work for everybody. Were you using these for an LSP or something? If so which one? And does this format work too? Or do we need to figure out how to adapt the type hints in a way that work for multiple systems?

Reply:
LDoc has quite a few ways of handing tags. Before I get too fancy with them I want to know if they are going to be used for anything other that `ldoc`.

For example `@tparam string var` defines a parameter of type string named var, but it also has `\string var` as an alias for the same thing. We can also setup custom types, which I want to do so we can have `@tparam measurement width` or `@tparam ast tree` for example that are linked to SILE.types.measurement and SILE.types.ast (or whatever that ends up going, right now it's just a table).

Reply:
@alerque There are a bunch of LSP and VSCode add-ons for Lua. I currently use https://github.com/Tencent/LuaHelper -- I can't remember why I picked that one... I also tried https://luals.github.io/ at one point (installed on my main host, but disabled).

Both have helpers to generate the documentation from code, and a fairly similar syntax for parametes, with `@param name type description` -- in line with my plugins for other languages (Typescript, Python...), mimicking their conventions for optional parameters and variant types (e.g. `table?` and/or `table|nil` etc.)

I am afraid LDoc's `@tparam` is quite specific to it, and doesn't have much support elsewhere... Likewise, marking optionals with`[opt]` seems unusual.
It's annoying people keep reinventing the wheel there, with their own language documentation things. 
I am not going to type those manually if I can avoid :) Hopefully, Copilot can help here too, though I don't always have it on (esp. in vacations, with a low/limited bandwidth).

Reply:
Addition: It seems both plugins I mentioned use EmmyLua annotations (or a variant?).

Reply:
:facepalm: :facepalm: :facepalm:

EmmyLua looks like a disaster to me, but the luals (formerly Sumneko Lua) stuff looks promising. They are not the same but both allow some customization and it might be possible to map out a dual-purpose ground that has some meaning to both.

https://github.com/lunarmodules/ldoc/issues/403

Reply:
Also see #1999.